loki/transformations/single_column/vector.py

```diff
@@ -450,6 +480,196 @@ def transform_subroutine(self, routine, **kwargs):
+        ``!$loki vector-reduction(<reduction clause>)``/
+        ``!$loki end vector-reduction`` is encountered, we replace
+        existing ``!$loki loop vector`` loop pragmas and add the
+        reduction keyword and clause. These will be turned into
+        OpenACC equivalents by :any:`SCCAnnotate`.
+        """
+        with pragma_regions_attached(routine):
+            for region in FindNodes(ir.PragmaRegion).visit(section):
+                if is_loki_pragma(region.pragma, starts_with='vector-reduction'):
+                    if (reduction_clause := re.search(r'reduction\([\w:0-9 \t]+\)', region.pragma.content)):
+
+                        loops = FindNodes(ir.Loop).visit(region)
+                        if loops:
+                            pragma = ir.Pragma(keyword='loki', content=f'loop vector {reduction_clause[0]}')
+                            # Update loop and region in place to remove marker pragmas
+                            loops[0]._update(pragma=(pragma,))
+                            region._update(pragma=None, pragma_post=None)
+
+    @staticmethod
+    def _get_loop_bound(bound, call_arg_map):
+        if isinstance(bound, tuple):
+            for alias in bound:
+                alias_arg = alias
+                elem = None
+                if '%' in alias:
+                    elem = alias_arg.split('%')[1]
+                    alias_arg = alias_arg.split('%')[0]
+                if alias_arg in call_arg_map:
+                    return (call_arg_map[alias_arg.lower()], elem)
+        return (call_arg_map[bound.lower()], None)
+
+    def transform_subroutine(self, routine, **kwargs):
+        """
+        Wrap vector-parallel sections in vector :any:`Loop` objects.
+        This wraps all thread-parallel sections within "kernel"
+        routines or within the parallel loops in "driver" routines.
+        The markers placed by :any:`SCCDevectorTransformation` are removed
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            Subroutine to apply this transformation to.
+        role : str
+            Must be either ``"kernel"`` or ``"driver"``
+        targets : tuple or str
+            Tuple of target routine names for determining "driver" loops
+        """
+        role = kwargs['role']
+        targets = tuple(str(t).lower() for t in as_tuple(kwargs.get('targets', None)))
+        # ignore = kwargs.get('ignore', ())
+        item = kwargs.get('item', None)
+        ignore = item.ignore if item else () + tuple(str(t).lower() for t in as_tuple(kwargs.get('ignore', None)))
+
+        if role == 'kernel':
+            # Skip if kernel is marked as `!$loki routine seq`
+            if check_routine_sequential(routine):
+                return
+
+            if self.horizontal.index not in routine.variables:
+                jl = get_integer_variable(routine, self.horizontal.index)
+                routine.arguments += (jl.clone(type=jl.type.clone(intent='in')),)
```

Comment:
This one is also not covered. I think that becomes relevant when you have a routine that only dispatches to other kernels but doesn't have any loops itself?